coalib/bearlib/abstractions/Linter.py

```diff
@@ -565,21 +591,82 @@ def run(self, filename, file, **kwargs):
 
-        def __repr__(self):
-            return '<{} linter object (wrapping {!r}) at {}>'.format(
-                type(self).__name__, self.get_executable(), hex(id(self)))
+    class GlobalLinterBase(LinterBase, GlobalBear):
+
+        @staticmethod
+        def create_arguments(config_file):
+            """
+            Creates the arguments for the linter.
 
+            You can provide additional keyword arguments and defaults. These
+            will be interpreted as required settings that need to be provided
+            through a coafile-section.
+
+            :param config_file:
+                The path of the config-file if used. ``None`` if unused.
+            :return:
+                A sequence of arguments to feed the linter-tool with.
+            """
+            raise NotImplementedError
+
+        def run(self, **kwargs):
+            # Get the **kwargs params to forward to `generate_config()`
+            # (from `_create_config()`).
+            generate_config_kwargs = FunctionMetadata.filter_parameters(
+                self._get_generate_config_metadata(), kwargs)
+
+            with self._create_config(
+                    **generate_config_kwargs) as config_file:
+                # And now retrieve the **kwargs for `create_arguments()`.
+                create_arguments_kwargs = (
+                    FunctionMetadata.filter_parameters(
+                        self._get_create_arguments_metadata(), kwargs))
+                args = self.create_arguments(config_file,
+                                             **create_arguments_kwargs)
+
+                try:
+                    args = tuple(args)
+                except TypeError:
+                    self.err('The given arguments '
+                             '{!r} are not iterable.'.format(args))
+                    return
+
+                arguments = (self.get_executable(),) + args
+                self.debug("Running '{}'".format(' '.join(arguments)))
+
+                output = run_shell_command(
+                    arguments,
+                    stdin=None,
+                    cwd=self.get_config_dir())
+
+                output = tuple(compress(
+                    output,
+                    (options['use_stdout'], options['use_stderr'])))
+                if len(output) == 1:
+                    output = output[0]
+
+                process_output_kwargs = FunctionMetadata.filter_parameters(
+                    self._get_process_output_metadata(), kwargs)
+                return self.process_output(output, **process_output_kwargs)
+
+    if options['global_bear']:
+        LinterBaseClass = GlobalLinterBase
+    else:
+        LinterBaseClass = LocalLinterBase
     # Mixin the linter into the user-defined interface, otherwise
     # `create_arguments` and other methods would be overridden by the
     # default version.
     result_klass = type(klass.__name__, (klass, LinterBase), {
```

Comment:
ok just looked at this side by side it's only a few file args here and there passed through, that can probably be hacked into the kwargs or so so we don't even need so many branches. Most certainly they can be joined.